Lua refactor

trws opened the pull request:
Reworking lua to allow easier substitution of the base lua implementation.  Also adding in a maintained version of luajit and re-factoring the entire stack to use a custom build_system to centralize functionlity like environment variable management and luarocks installation.

Requires: #28853

trws committed:
extensions: allow multiple "extends" directives

This will allow multiple extends directives in a package as long as only one of
them is selected as a dependency in the concrete spec.

trws committed:
lua: refactor lua implementation packages

* add LuaImplPackage as a central place to manage common state for all
  lua implementations
* move luarocks installation and environment setup into the package type
  used for all luas
* refactor lua and lua-luajit to use the new package type

trws committed:
add maintained luajit-openresty package

trws committed:
add base package type for lua extension packages

trws committed:
update several lua packages and pkgkit to use new package types

trws committed:
isort applied

trws committed:
fix mypy checks for directives

trws committed:
address style errors

adamjstewart reviewed (commented):
Haven't looked at this in too much detail yet. My first impression is that I agree that we need a `LuaPackage` base class for Lua extensions. I'm less certain about `LuaImplPackage` though. This seems to relate in many ways to #17475, and I think a lot of the "hackiness" I feel from this PR is due to the fact that 1) we can't directly extend a virtual package, and 2) we don't want to copy-n-paste all the activation logic to multiple Lua providers. Python and Java face these same issues, and we haven't found a good solution yet. It would be good to discuss if this solution is something that we want to use in general or if there's a better solution that can be used for all virtual providers/extensions.

Also, once we decide on a final design for this, we'll need docs in `lib/spack/docs/build_systems/luapackage.rst` and possibly build-system detection in `lib/spack/spack/cmd/create.py`.

trws commented:
Yup, you're absolutely right that it relates to #17475.  If we could extend a virtual, then the extends(when) stuff could all come out.

I decided to go this way for lua specifically because:
* all of the implementations should use the same logic for:
  * extension
  * build and run environment setup
  * luarocks installation
  * providing access to luarocks

That is a whole lot of code that ought to be implemented in only one place, and at the moment the only way I know to do that within spack is to make it a build system.  I argued with myself a bit about making all of that logic part of an actual lua-lang *package* that would depend on one of the three depending on variants or something, but that seems even worse, or making it a package that gets imported by the others and used as a base, but again that didn't seem better.

What I'm trying to get to I suppose is that there are two problems here:
* no good way to extend virtuals (as in #17475)
* no good way to share common code between related packages other than build systems

The latter is something I want to find a way to do for packages like LLVM and the ROCM family.  It bugs the crap out of me that the same set of version logic is copied into each and every rocm package, so every time there's a new version every single one has to get touched to add a new number in the exact same list.  This is part of why LLVM ended up as a single package again instead of split out too, it should be several packages with *one central implementation and version*.  Not sure how best to fix that, but given the mechanisms available this seemed least hacktastic.

The thing I still really don't like is that luarocks is embedded in lua.  I say that despite being the person that did that in the first place.  It almost has to be there because that's where things should activate to, but at the same time it should be able to have a version independent of lua.

Anyway, sorry for the book, 100% agree there are things to resolve here.

adamjstewart commented:
I've also thought about whether it makes sense for virtual dependencies to require actual `package.py` files so that this kind of code can be shared and we don't need to parse every single `package.py` just to determine which packages provide what.

trws commented:
That might make for a nice solution.  By far the worst part of implementing this refactor was getting the two build_system files to work without import cycles in bootstrap.  Using normal package directives in build_system classes is *very* fragile.

adamjstewart commented:
Pinging @alalazo @becker33 @tgamblin on this one because they usually enjoy brainstorming these kinds of conceptual design questions. See #17475 for a good description of at least half of the issue. The issue of reusing the same code for all virtual providers is kind of a second issue that we should also think about.

trws commented:
Anybody have opinions: @alalazo, @becker33, @tgamblin?  It's a bit painful to be stuck with upstream lua only without something like this patch.  I don't mind re-working if we want a different mechanism, but some way forward would be really helpful.

trws committed:
fix vermin error

adamjstewart commented:
Took another look at this and I think my thoughts on this PR are mostly the same as when I last looked at it. Basically, `LuaPackage` looks good to me, but `LuaImplPackage` feels like a hack to work around other deficiencies in Spack.

I think the "right" thing to do is to fix #17475. Of course, this is easier said than done, and might take a while to get right. We might have to merge something like this PR while we wait.

Another design question I have is what actually belongs in `lib/spack/spack/build_systems`? Originally it was only build systems, then we started adding mixins for things like Cuda and ROCm. Obviously mixins are not build systems, and `LuaImplPackage` is even more different. I think `LuaImplPackage` may be better to store in the `package.py` of one of the Lua providers. For example, we do something very similar with `py-pillow` and `py-pillow-simd`. Thoughts?

trws committed:
move LuaImplPackage into the lua package

alalazo commented on lib/spack/spack/build_systems/lua.py:
```diff
@@ -0,0 +1,59 @@
+
+    @property
+    def lua(self):
+        return Executable(join_path(self.spec['lua-lang'].prefix.bin, 'luarocks'))
```
Is this a typo:
```suggestion
        return Executable(join_path(self.spec['lua-lang'].prefix.bin, 'lua'))
```
?

trws replied:
Yes, that's a typo, thanks!

alalazo reviewed (changes requested):
I don't mind this implementation while we look at a better refactoring. For the `LuaImplPackage`, is there one of the three implementations that is a de-facto reference? If so we can put that class in the corresponding package and have other implementations:
```python
from spack.pkg.builtin.*** import LuaImplPackage
```

Minor: can you fix style issues, so we can get other checks running too?

trws committed:
refactor and fix style

trws committed:
fix typo in lua property

trws committed:
Add lua and luarocks properties to implementations

trws commented:
Ok, re-factored to put `LuaImplPackage` in the lua package.  Fixing a couple of style checking issues, and repaired the properties.  I'm going to go over and rebase the neovim fixup on top of this to get some better downstream testing.

adamjstewart commented:
> Also, once we decide on a final design for this, we'll need docs in `lib/spack/docs/build_systems/luapackage.rst` and possibly build-system detection in `lib/spack/spack/cmd/create.py`.

Ping on this. At this point you're probably the only Spack developer who has a good understanding of Lua's build system.

trws committed:
switch to luarocks make to avoid luarocks attempting to fetch dependencies

trws commented:
Ah, good point, I'll do that shortly.

trws committed:
tweak naming in LuaPackage and add create support

trws committed:
document LuaPackage

trws committed:
hook luapackage.rst into the toctree

trws commented:
The create integration is in and working.  I don't often work with rst, so I'm not as confident the docs are the way they need to be, but they are there.

trws committed:
fix rst syntax

adamjstewart commented on lib/spack/docs/build_systems/luapackage.rst:
```diff
@@ -0,0 +1,82 @@
+.. _luapackage:
+
+------------
+MesonPackage
```
This first section still references Meson

adamjstewart commented on lib/spack/docs/build_systems/luapackage.rst:
```diff
@@ -0,0 +1,82 @@
+^^^^^^^^^^^^^^^
+
+Packages that use the Lua/LuaRocks build system can be identified by the
+presence of a ``.*\.rockspec`` file. This file declares things
```
Probably better to use a glob `*.rockspec`

adamjstewart commented on lib/spack/docs/build_systems/luapackage.rst:
```diff
@@ -0,0 +1,82 @@
+presence of a ``.*\.rockspec`` file. This file declares things
+like build instructions and dependencies.
+
+It is common for the rockspec file to list the lua version required in
```
Can you add a 1-line example of this? This helps people know what text to search for in that file

trws replied:
Sure, I'll use this one:

```lua
dependencies = {
   "lua >= 5.1",
}
```

adamjstewart commented on lib/spack/spack/build_systems/lua.py:
```diff
@@ -0,0 +1,59 @@
+class LuaPackage(PackageBase):
+    """Specialized class for lua packages
+
+    Note that there is also a LuaRocksPackage for the common case where luarocks is the
```
Where is LuaRocksPackage?

trws replied:
Dang, that's a holdout from a refactor.  I initially had luapackage as a specialized MakefilePackage, but it just isn't consistent enough to be useful, so I got rid of it and promoted LuaRocksPackage to LuaPackage.

adamjstewart commented on lib/spack/spack/build_systems/lua.py:
```diff
@@ -0,0 +1,59 @@
+    extends('lua-luajit-openresty', when='^lua-luajit-openresty')
+    depends_on('lua-luajit-openresty+lualinks', when='^lua-luajit-openresty')
+
+    rocks_make_flags = []  # type: List[Text]
```
Probably better to make this a function like `cmake_args` or `configure_args`

trws replied:
Fair, updated.

adamjstewart commented on lib/spack/spack/cmd/create.py:
```diff
@@ -668,6 +679,7 @@ def __call__(self, stage, url):
             (r'/Rakefile$',           'ruby'),
             (r'/setup\.rb$',          'ruby'),
             (r'/.*\.pro$',            'qmake'),
+            (r'/.*\.rockspec$',       'lua'),
```
You can also update `lib/spack/spack/test/build_system_guess.py` to add tests for this.

trws replied:
Thanks for that, added.

trws committed:
address comments

trws committed:
fix rst code-block syntax

adamjstewart commented on lib/spack/docs/build_systems/luapackage.rst:
```diff
@@ -0,0 +1,93 @@
+like build instructions and dependencies.
+
+It is common for the rockspec file to list the lua version required in
+a dependency, the LuaPackage class adds appropriate dependencies on a Lua
```
```suggestion
a dependency. The LuaPackage class adds appropriate dependencies on a Lua
```

adamjstewart commented on lib/spack/docs/build_systems/luapackage.rst:
```diff
@@ -0,0 +1,93 @@
+
+   dependencies = {
+      "lua >= 5.1",
+   }
```
What would be the equivalent syntax for Spack? Does we need to specify it for multiple possible providers at the moment?

trws replied:
I suppose it would be something like:

```python
depends_on('lua@5.1:5.1.99', when='^lua')
```

It's only one because it's 5.1, and all the luajit variants are 5.1 variants.  If it were 5.3, only lua would be valid, so it would become:

```python
depends_on('lua@5.3')
```

That's actually a bit of an oddity with all this.  I considered putting a version on lua-lang, from a logical standpoint it should probably be there, but there's only one lua implementation that provides anything other than 5.1.

trws replied:
I added some stuff to the docs, but in principle this should now be `depends_on('lua-lang@5.1:')`, as long as the package doesn't care what implementation it gets.

adamjstewart commented on lib/spack/spack/build_systems/lua.py:
```diff
@@ -0,0 +1,60 @@
+
+from llnl.util.filesystem import find, join_path
+
+from spack.directives import *
```
Let's avoid star imports and specify depends_on, extends explicitly

adamjstewart commented on lib/spack/spack/build_systems/lua.py:
```diff
@@ -0,0 +1,60 @@
+class LuaPackage(PackageBase):
+    """Specialized class for lua packages
+
+    Note that there is also a LuaPackage for the common case where luarocks is the
+    build system"""
```
I think this line can be deleted

adamjstewart commented on lib/spack/spack/build_systems/lua.py:
```diff
@@ -0,0 +1,60 @@
+    depends_on('lua-luajit-openresty+lualinks', when='^lua-luajit-openresty')
+
+    def preprocess(self, spec, prefix):
+        """Extend this to preprocess source before building with luarocks"""
```
```suggestion
        """Override this to preprocess source before building with luarocks"""
```

adamjstewart commented on lib/spack/spack/build_systems/lua.py:
```diff
@@ -0,0 +1,60 @@
+
+    @property
+    def lua(self):
+        return Executable(join_path(self.spec['lua-lang'].prefix.bin, 'lua'))
```
```suggestion
        return Executable(self.spec['lua-lang'].prefix.bin.lua)
```

adamjstewart commented on lib/spack/spack/build_systems/lua.py:
```diff
@@ -0,0 +1,60 @@
+
+    @property
+    def luarocks(self):
+        return Executable(join_path(self.spec['lua-lang'].prefix.bin, 'luarocks'))
```
```suggestion
        return Executable(self.spec['lua-lang'].prefix.bin.luarocks)
```

adamjstewart commented on lib/spack/spack/build_systems/lua.py:
```diff
@@ -0,0 +1,60 @@
+            specs = find('.', '*.rockspec', recursive=False)
+            if specs:
+                rock = specs[0]
+        rocks_args = list(self.luarocks_args())
```
Isn't this already a list?

trws replied:
That was there to copy the list more than convert it, but now that it's returned from a function it's no longer necessary.  Removed.

adamjstewart commented on lib/spack/docs/build_systems/luapackage.rst:
```diff
@@ -0,0 +1,93 @@
+
+The ``LuaPackage`` base class comes with the following phases:
+
+#. ``preprocess`` - generate ninja files
```
Are there any examples of packages that use this (or will need this) phase? Or is this just hypothetical?

trws replied:
I had one that needed it, but reworked it using a newer luarocks dependency.  There are a *lot* of old largely unmaintained but critical lua packages around, I expect it to come up every so often.

trws replied:
This is actually back in active use, the lua-bitlib package needs it to fix the `luaL_reg` calls to be capitalized correctly.

adamjstewart commented on lib/spack/spack/build_systems/lua.py:
```diff
@@ -0,0 +1,60 @@
+
+
+import os
+from typing import List, Text
```
I think these can be removed now

adamjstewart commented on lib/spack/spack/cmd/create.py:
```diff
@@ -187,6 +187,19 @@ def cmake_args(self):
+
+    body_def = """\
+    def luarocks_args(self):
+        # FIXME: Add arguments to `luarocks make` other rockspec path
```
```suggestion
        # FIXME: Add arguments to `luarocks make` other than rockspec path
```

adamjstewart commented on var/spack/repos/builtin/packages/lua-lpeg/package.py:
```diff
@@ -16,8 +16,7 @@ class LuaLpeg(Package):
-    depends_on("lua@:5.1.9", when="@:0.12.1")
+    depends_on("lua@:5.1.9", when="@:0.12.1^lua")
 
     def install(self, spec, prefix):
```
Does the default install method not work?

trws replied:
Thanks for the reminder.  This version of the package would not have worked, no, because lpeg was using the actual rock archive from the luarocks distro rather than a source package.  Basically a luarocks wheel.  I'm reworking it to use the source instead then the default works just fine.

trws replied:
Spack now supports the rock archives directly, and this is reworked to use it.

adamjstewart commented on var/spack/repos/builtin/packages/lua/package.py:
```diff
@@ -3,14 +3,219 @@
-from spack import *
+import spack.build_environment
+from spack.build_systems.makefile import MakefilePackage
+from spack.directives import *
```
This shouldn't be needed in packages, these are all already included

trws replied:
True, same for the filesystem ones.  They came in from being needed in the build_system version, but are now gone.

adamjstewart commented on var/spack/repos/builtin/packages/lua/package.py:
```diff
@@ -3,14 +3,219 @@
+import spack.build_environment
+from spack.build_systems.makefile import MakefilePackage
+from spack.directives import *
+from spack.pkgkit import *
```
I think this is something we've been meaning to add to all packages at some point, not sure if we ever made any progress on that

trws replied:
Yeah, it's on my list.  That and getting things out of the module, and into package instances or similar so we can actually typecheck these darned things would make me really happy.  It keeps getting pushed back by functional stuff for build though.  At least putting it in as I go gets most of the type checking failures out for the ones I'm touching.

adamjstewart replied:
If we could get it in before 0.18 (sometime in the next week) that would be amazing. It's hard to add after because it becomes difficult to backport.

trws replied:
I can try to dig out my old rope script and just do it, it basically can't break anything, so I think it's mainly a question of getting it in and writing something to make sure every new package has it.  Maybe add an isort rule? Will look into it.

adamjstewart commented on var/spack/repos/builtin/packages/lua/package.py:
```diff
@@ -3,14 +3,219 @@
+
+    @property
+    def lua(self):
+        return Executable(join_path(self.spec.prefix.bin, 'lua'))
```
```suggestion
        return Executable(self.spec.prefix.bin.lua)
```

adamjstewart commented on var/spack/repos/builtin/packages/lua/package.py:
```diff
@@ -3,14 +3,219 @@
+
+    @property
+    def luarocks(self):
+        return Executable(join_path(self.spec.prefix.bin, 'luarocks'))
```
```suggestion
        return Executable(self.spec.prefix.bin.luarocks)
```

adamjstewart commented on var/spack/repos/builtin/packages/lua/package.py:
```diff
@@ -3,14 +3,219 @@
+
+        In most cases, extensions will only need to have two lines::
+
+        luarocks('--tree=' + prefix, 'install', rock_spec_path)
```
I think this line needs to be indented for the double colon above to work. Not that we generate docs for packages anyway.

adamjstewart commented on var/spack/repos/builtin/packages/lua/package.py:
```diff
@@ -3,14 +3,219 @@
+        luarocks('--tree=' + prefix, 'install', rock_spec_path)
+        """
+        # Lua extension builds can have lua and luarocks executable functions
+        module.lua = Executable(join_path(self.spec.prefix.bin, "lua"))
```
```suggestion
        module.lua = Executable(self.spec.prefix.bin.lua)
```

adamjstewart commented on var/spack/repos/builtin/packages/lua/package.py:
```diff
@@ -3,14 +3,219 @@
+        """
+        # Lua extension builds can have lua and luarocks executable functions
+        module.lua = Executable(join_path(self.spec.prefix.bin, "lua"))
+        module.luarocks = Executable(join_path(self.spec.prefix.bin, "luarocks"))
```
```suggestion
        module.luarocks = Executable(self.spec.prefix.bin.luarocks)
```

trws committed:
Update lib/spack/docs/build_systems/luapackage.rst

Co-authored-by: Adam J. Stewart <ajstewart426@gmail.com>

trws committed:
Apply suggestions from code review

Co-authored-by: Adam J. Stewart <ajstewart426@gmail.com>

trws committed:
Apply suggestions from code review

Co-authored-by: Adam J. Stewart <ajstewart426@gmail.com>

trws committed:
resolve comments

trws committed:
gerge branch 'lua-refactor' of github.com:trws/spack into lua-refactor

trws committed:
support .src.rock archives, update lpeg and docs

This is a bit of a big change.  Turns out lpeg does not provide a
reasonably buildable source package upstream, so I built out support for
luarocks.org links including auto-building them, and finding,
checksumming and fetching them appropriately.  This is slightly
dangerous because luarocks can download dependencies under us if someone
writes a package the right way, but it's the best way to support very
important but funky packages like this.

Additionally, the `lua-lang` virtual is now versioned so that a package
that requires Lua 5.1 semantics can get any lua, but one that requires
5.2 will only get upstream lua.  May be worth considering using compat53
to provide 5.2 and 5.3 on top of luajit, but that's a lot more work than
I think should go into this right now.

trws committed:
use new lua package support in lua-luafilesystem

trws committed:
update bitlib package, rework phases to unpack

Reworked phases and redocumented to support preprocessing inside
unpacked rock source directory.  The lua-bitlib package requires this to
build.

trws committed:
add source rock support to create.py

trws committed:
allow for luajit installing a versioned link

trws committed:
allow luarocks to find deps, fix luaposix

 This ended up being a lot more than I meant it to be.  The luaposix
 package requires lua-bit32, but only when built with a lua conforming
 to version 5.1.  This adds the package, and the dependencies, but
 exposed a problem with luarocks dependency detection.  Since we're
 installing each package in its own "tree" and there's no environment
 variable to list extra trees, spack now generates a luarocks config
 file that lists all the trees of all the dependencies, and references
 it by setting `LUAROCKS_CONFIG` in the build environment of every
 LuaPackage.  This allows luarocks to find the spack installed
 dependencies correctly rather than trying (and failing) to download
 them.

trws committed:
remove debug prints

trws committed:
fix style issues

trws commented:
Ok, that took a *lot* more changes than I had been hoping, but I updated every lua package I could find that was manually using luarocks, and extended the luarocks support to correctly handle cross-tree dependencies and source rock archives, along with new support in create and fetch to handle that.  😮‍💨 In the end, it seems to work, across lua implementations and versions and packages.  Also fixed missing dependencies we were accidentally fetching from the internet previously.

trws committed:
allow concretizer to resolve virtual versions

This allows the luaposix dependence on bit32 to be expressed correctly.
Docs change coming shortly, massively simplifies writing version deps
for lua.

trws commented:
Thanks to @tgamblin for some help spelunking the ASP, the last commit fixes concretizing to handle `when='variant@version'` allowing the luaposix bit32 dependency to be described accurately.  Assuming this passes, it should be ready for a re-review.

adamjstewart commented:
One alternative to the `unpack` phase of `LuaPackage` would be to support directly extracting `*.rock` files in `lib/spack/spack/util/compression.py`. Advantage would be one less phase and chdir and removes the need for `extract=False` in the packages. Disadvantage would be needing to search for `luarocks` in the `PATH` via `which` in this location, but that shouldn't be an issue since we'll always have a dep on lua.

This PR is starting to get big. We may want to split this into multiple PRs if it gets much bigger.

trws commented:
I actually tried to implement it that way first, but ran into issues with that solution because the unpack makes a two-level directory result that doesn't get handled properly by the directory identification machinery, and honestly I don't even know where to look to fix that.  If you can point me in the general direction, I wouldn't mind doing it, but it was becoming too much work, hence the rework into a phase.

adamjstewart reviewed (dismissed):
I'm happy with the current state of this. We should probably get a review from at least one more person before merging. Great work!

alalazo commented:
With respect to this: https://github.com/spack/spack/pull/28854#issuecomment-1112775790 did we check if there's any impact on concretization performance?

alalazo commented on lib/spack/spack/build_systems/lua.py:
```diff
@@ -0,0 +1,99 @@
+    list_depth = 1  # LuaRocks requires at least one level of spidering to find versions
+    depends_on('lua-lang')
+    extends('lua', when='^lua')
+    extends('lua-luajit', when='^lua-luajit')
+    depends_on('luajit', when='^lua-luajit')
+    depends_on('luajit', when='^lua-luajit-openresty')
+    depends_on('lua-luajit+lualinks', when='^lua-luajit')
+    extends('lua-luajit-openresty', when='^lua-luajit-openresty')
+    depends_on('lua-luajit-openresty+lualinks', when='^lua-luajit-openresty')
```
I think this should work too and is a bit more readable:
```suggestion
    with when('^lua-luajit'):
        extends('lua-luajit')
        depends_on('luajit')
        depends_on('lua-luajit+lualinks')
    
    with when('^lua-luajit-openresty'):
        extends('lua-luajit-openresty')
        depends_on('luajit')
        depends_on('lua-luajit-openresty+lualinks')
```

alalazo commented on lib/spack/spack/build_systems/lua.py:
```diff
@@ -0,0 +1,99 @@
+        if os.path.splitext(self.stage.archive_file)[1] == '.rock':
+            directory = self.luarocks('unpack', self.stage.archive_file, output=str)
+            dirlines = directory.split('\n')
+            os.chdir(dirlines[2])
```
Can we store the directory and then use the context manager `working_dir` where appropriate? That should restore the current working directory on exit from the context.

trws replied:
We can certainly store the directory, but I've had zero luck figuring out how to use `with working_dir` in an un-scoped manner like this needs.  Is it possible to set the `build_directory` dynamically in an early phase and tell spack to re-evaluate it or something? I agree that I'm unfond of this but hadn't come up with a better solve.

alalazo replied:
If that is hard to do, I would not block merging this PR. Can you leave a `TODO` comment though or something like that?

trws replied:
Sure, adding one now.

alalazo commented on lib/spack/spack/build_systems/lua.py:
```diff
@@ -0,0 +1,99 @@
+    def setup_build_environment(self, env):
+        env.set('LUAROCKS_CONFIG', self._luarocks_config_path())
+
+    def preprocess(self, spec, prefix):
+        """Override this to preprocess source before building with luarocks"""
+        pass
```
Why are we adding an empty phase with a hook that runs before that (`_generate_luarocks_config`)? Wouldn't make more sense to have the hook be the phase and eventually have people `run_after` that?

trws replied:
The config generation could be a phase. I think I did it this way so that if the user uses luarocks for something during preprocess they'll get the config, they could edit the generated config in preprocess, but they can't accidentally prevent it from being run by overloading the phase.  The empty phase is there because there are a number of packages that need editing or patching to compile, serves a function much like `edit` in MakefilePackage.

alalazo replied:
> The config generation could be a phase.

Yeah, let's make it a phase on its own, I'd be good with that.

alalazo commented on lib/spack/spack/package.py:
```diff
@@ -1324,7 +1324,7 @@ def extendee_spec(self):
 
         # If the extendee is in the spec's deps already, return that.
-        for dep in self.spec.traverse(deptype=('link', 'run')):
+        for dep in self.spec.traverse(deptypes=('link', 'run')):
```
This introduces a bug. It's weird, but the argument is `deptype`, see #29397

trws replied:
I wonder why I did this... I have a vague memory that it actually fixed something, which is disturbing in its own way, I'll look into it.

trws committed:
Update lib/spack/spack/build_systems/lua.py

Co-authored-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo commented on lib/spack/spack/build_systems/lua.py:
```diff
@@ -0,0 +1,102 @@
+
+from llnl.util.filesystem import find
+
+from spack.directives import depends_on, extends
```
```suggestion
from spack.directives import depends_on, extends
from spack.multimethod import when
```

alalazo commented on lib/spack/spack/build_systems/lua.py:
```diff
@@ -0,0 +1,102 @@
+        extends('lua-luajit')
+        depends_on('luajit')
+        depends_on('lua-luajit+lualinks')
+    
```
```suggestion

```

alalazo reviewed (commented):
A few fixes for style

alalazo committed:
Fix missing import and style issue

trws committed:
convert generate to a phase and fix package

trws committed:
style fixes

trws commented:
> With respect to this: [#28854 (comment)](https://github.com/spack/spack/pull/28854#issuecomment-1112775790) did we check if there's any impact on concretization performance?

I did some testing, both on the full PR and on just the changes in the ASP.  The change in the ASP was basically nil, but with the full PR there's a small but noticeable increase in solve time.  I'm guessing it's because of the additional virtual.

trws commented:
@spackbot run pipeline

alalazo reviewed (approved):
Thanks!
